pallets/loans/src/tests/market.rs

```diff
@@ -0,0 +1,187 @@
+            );
+            // base_rate greater than jump_rate
+            assert_noop!(
+                Loans::$call(Origin::signed(ALICE), DOT, {
+                    let mut market = MARKET_MOCK;
+                    market.rate_model = InterestRateModel::new_jump_model(
+                        Rate::saturating_from_rational(10, 100),
+                        Rate::saturating_from_rational(9, 100),
+                        Rate::saturating_from_rational(14, 100),
+                        Ratio::from_percent(80),
+                    );
+                    market
+                }),
+                Error::<Runtime>::InvalidRateModelParam
+            );
+            // jump_rate greater than full_rate
+            assert_noop!(
+                Loans::$call(Origin::signed(ALICE), DOT, {
+                    let mut market = MARKET_MOCK;
+                    market.rate_model = InterestRateModel::new_jump_model(
+                        Rate::saturating_from_rational(5, 100),
+                        Rate::saturating_from_rational(15, 100),
+                        Rate::saturating_from_rational(14, 100),
+                        Ratio::from_percent(80),
+                    );
+                    market
+                }),
+                Error::<Runtime>::InvalidRateModelParam
+            );
+        })
+    };
+}
+
+#[test]
+fn active_market_sets_state_to_active() {
+    ExtBuilder::default().build().execute_with(|| {
+        Loans::add_market(Origin::signed(ALICE), NATIVE, PENDING_MARKET_MOCK).unwrap();
+        assert_eq!(Loans::market(&NATIVE).unwrap().state, MarketState::Pending);
+        Loans::active_market(Origin::signed(ALICE), NATIVE).unwrap();
+        assert_eq!(Loans::market(&NATIVE).unwrap().state, MarketState::Active);
+    })
+}
+
+#[test]
+fn active_market_does_not_modify_unknown_market_currencies() {
+    ExtBuilder::default().build().execute_with(|| {
+        assert_noop!(
+            Loans::active_market(Origin::signed(ALICE), NATIVE),
+            Error::<Runtime>::MarketDoesNotExist
+        );
+    })
+}
+
+#[test]
+fn add_market_does_not_insert_a_new_market_with_duplicated_currencies() {
+    ExtBuilder::default().build().execute_with(|| {
+        assert_noop!(
+            Loans::add_market(Origin::signed(ALICE), DOT, MARKET_MOCK),
+            Error::<Runtime>::MarketCurrencyAlreadyExists
+        );
+    })
+}
+
+#[test]
+fn add_market_ensures_that_market_state_must_be_pending() {
+    ExtBuilder::default().build().execute_with(|| {
+        Loans::add_market(Origin::signed(ALICE), NATIVE, MARKET_MOCK).unwrap();
+        assert_eq!(Loans::market(&NATIVE).unwrap().state, MarketState::Pending);
+    })
+}
+
+#[test]
+fn add_market_has_sanity_checks_for_rate_models() {
+    rate_model_sanity_check!(add_market);
+}
+
+#[test]
+fn add_market_successfully_stores_a_new_market() {
+    ExtBuilder::default().build().execute_with(|| {
+        assert!(Loans::market(&NATIVE).is_err());
```

Comment:
Handle the specific error MarketDoesNotExist

Reply:
Ok